slycot/math.py

```diff
@@ -70,6 +70,361 @@ def mc01td(dico,dp,p):
     return out[:-2]
 
 
+def mb03vd(n, ilo, ihi, A):
+    """ HQ, Tau = mb03vd(n, ilo, ihi, A)
+
+    To reduce a product of p real general matrices A = A_1*A_2*...*A_p
+    to upper Hessenberg form, H = H_1*H_2*...*H_p, where H_1 is
+    upper Hessenberg, and H_2, ..., H_p are upper triangular, by using
+    orthogonal similarity transformations on A,
+
+            Q_1' * A_1 * Q_2 = H_1,
+            Q_2' * A_2 * Q_3 = H_2,
+                   ...
+            Q_p' * A_p * Q_1 = H_p.
+
+    Parameters
+    ----------
+
+    n : int
+            The order of the square matrices A_1, A_2, ..., A_p.
+            n >= 0.
+
+    ilo, ihi : int
+            It is assumed that all matrices A_j, j = 2, ..., p, are
+            already upper triangular in rows and columns [:ilo] and
+            [ihi:n], and A_1 is upper Hessenberg in rows and columns
+            [:ilo] and [ihi:n], with A_1[ilo-1,ilo] = 0 (unless
```

Comment:
This paragraph is fairly tough going.  Is the idea that the upper and lower blocks `A[:ilo,:ilo]` and `A[ihi:n, ihi:n]` are diagonal (or Hessenberg) already?

Is `A_1[ilo-1, ilo] = 0` part of the Hessenberg condition (in which case, shouldn't it be `A_1[ilo, ilo-2] = 0` ?), or is this an extra condition?

Reply:
This comes from the fact that Python is 0-indexed while the Fortran SLICOT routines count from 1. My request for review is to check if I made the conversion correctly.

Reply:
the comment header from `MB03VD.f` is:

```
C             already upper triangular in rows and columns 1:ILO-1 and
C             IHI+1:N, and A_1 is upper Hessenberg in rows and columns
C             1:ILO-1 and IHI+1:N, with A_1(ILO,ILO-1) = 0 (unless
C             ILO = 1), and A_1(IHI+1,IHI) = 0 (unless IHI = N).
C             If this is not the case, ILO and IHI should be set to 1
C             and N, respectively.
C             1 <= ILO <= max(1,N); min(ILO,N) <= IHI <= N.
```

If this is Matlab-type indexing, which I'd guess it is (otherwise I think `IHI:N` would read `IHI:N+1`), then it's inclusive of end-points (closed), while Python is half-open.  For `ILO=5`, Matlab 1:ILO-1 means [1,2,3,4]; Python [:ilo] is [0,1,2,3,4], while we want [0,1,2,3].  So should that be "upper diagonal (or Hessenberg) in `[:ilo-1]`" ? This also makes sense for the lower bound of 1 on ilo; for ilo = 1 the first `[:ilo-1] = [:0] #=> empty` rows are upper diagonal (or Hessenberg).

I think the `ihi` expressions are correct: suppose N=10, IHI=7.  Fortran: `IHI+1:N = 8:10 = [8,9,10]`; Python: `[ihi:n] = 7,8,9`.

I still think the `A[ilo-1, ilo] = 0` comment is wrong, though I see it is straight from the SLICOT comment.  I'll add another comment later on this.

Reply:
N = 5; ilo = 2

A(ILO, ILO-1) = A(2,1) = 0

  1 2 3 4 5
1 d u u u u
2 x d u u u
3 x 0 d u u
4 x x x d u
5 x x x x d

I got this confused -- I thought the specified zero was above the diagonal.  It looks like this is an additional restriction on  A_1, i.e., beyond being only upper Hessenberg.

The Fortranesque `A_1(ILO,ILO-1) = 0` and `A_1(IHI+1,IHI) = 0` map correctly to the Pythonesque `A_1[ilo-1,ilo] = 0` and  `A_1[ihi,ihi-1] = 0`.

Reply:
So you now agree that `A[ilo-1, ilo] = 0` is correct?

Reply:
Yes, that seems right.

Reply:
Let's leave MATLAB out of here and focus on Fortran vs Python ;)

Fortran `(1,ILO-1)` has ILO-1 entries, so you're right it should be Python `[:ilo-1]`. Still confused about the `:n` now. Need another minute to think about it.

Reply:
A[ilo-1, ilo] = 0 is NOT correct. Shouldn't it be A[ilo-1:ilo-2]?

Reply:
The original Fortran comment in MB03VD.f line 51 is `with A_1(ILO,ILO-1) = 0`.  I think this specifies a single entry at (1-based) row `ILO-1` and column `ILO`, not a range (`,`, not ':').  This mean `A[ilo-1, ilo] = 0` is OK?

I think this diff is still needed for `mb03vd`

```
diff --git a/slycot/math.py b/slycot/math.py
index 39a7ea7..41b8fa1 100644
--- a/slycot/math.py
+++ b/slycot/math.py
@@ -92,9 +92,9 @@ def mb03vd(n, ilo, ihi, A):
 
     ilo, ihi : int
             It is assumed that all matrices A_j, j = 2, ..., p, are
-            already upper triangular in rows and columns [:ilo] and
+            already upper triangular in rows and columns [:ilo-1] and
             [ihi:n], and A_1 is upper Hessenberg in rows and columns
-            [:ilo-1] and [ihi:n], with A_1[ilo-1,ilo-2] = 0 (unless
+            [:ilo-1] and [ihi:n], with A_1[ilo,ilo-1] = 0 (unless
             ilo = 1), and A_1[ihi,ihi-1] = 0 (unless ihi = n).
             If this is not the case, ilo and ihi should be set to 1
             and n, respectively.
        1 <= ilo <= max(1,n); min(ilo,n) <= ihi <= n.
```

Reply:
Yes it is a single entry, not a slice. Thus, 
`FortranEntry(A_1(ILO, ILO-1)) == PythonEntry(A_1(ilo-1, ilo-2))`

On [:ilo-1], i agree. Commited.